### DISCUSSION TITLE
url_for() is broken with multiple same blueprint

### DISCUSSION THREAD

**Discussion by sergiokessler** [pallets/flask#4378] [Q&A]
my situation is something like this:

`api.py`
```python
def blueprint_maker(bpname, db_session):

    api = Blueprint(bpname, __name__)
    [...]
    @api.route('/ping', methods=['GET'])
    def ping():
        response = {
            'data': [{
                'echo': 'echo!',
                'source': url_for('api.ping')
            }]
        }
        return jsonify(response)

    return api
```

and then:

`main.py`
```python
[snip]
ar = blueprint_maker('ar', SessionLocal_ar)
uy = blueprint_maker('uy', SessionLocal_uy)
br = blueprint_maker('uy', SessionLocal_br)
pt = blueprint_maker('uy', SessionLocal_pt)
[...]


app.register_blueprint(ar, url_prefix='/ar')
app.register_blueprint(uy, url_prefix='/uy')
app.register_blueprint(ar, url_prefix='/br')
app.register_blueprint(uy, url_prefix='/pt')
[...]
```

I expected `url_for()` to just work in this scenario, but instead, it gives me:

```
Could not build url for endpoint 'api.ping'. Did you mean 'ar.ping' instead?
```

isn't this supposed to work?

Environment:

- Python version: 3.9
- Flask version: 2

**Comment by davidism** [comment]
As the error says, you need to specify the URL you want to build. The url takes the form `bp_name.view_name`. The bp name is `ar`, `uy`, etc.

**Comment by sergiokessler** [comment]
Maybe I was not clear, but there is ONE api.py, and at that point I don't know how the name will be later...

I tried 
```python
url_for(f'{bpname}.ping')
```

but that didn't work (it just gives me the relative url, not the full one)

**Reply by tachyondecay** [reply]
Call `url_for()` with `_external=True` to generate an absolute URL.

Also note you can pass a `name` argument to `register_blueprint` to register the same blueprint with multiple unique names ([more info](https://flask.palletsprojects.com/en/2.0.x/api/#flask.Flask.register_blueprint)).

**Reply by sergiokessler** [reply]
thanks!

this is the code to make it work:
(note that I can't hardcode the blueprint `name` here)

```python
def blueprint_maker(bpname, db_session):

    api = Blueprint(bpname, __name__)
    [... do stuff with db_session ...]
    @api.route('/ping', methods=['GET'])
    def ping():
        response = {
            'data': [{
                'echo': 'echo!',
                'source': url_for(f'{bpname}.ping', _external=True)
            }]
        }
        return jsonify(response)

    return api
```

(and yes, I'm using the `name` argument to register the same bp multiple times, but the _bp implementation_ does not know what will be those names)
